rust/lance-io/src/scheduler.rs

```diff
@@ -3,42 +3,238 @@
 
 use bytes::Bytes;
 use futures::channel::oneshot;
-use futures::stream::FuturesUnordered;
-use futures::{FutureExt, StreamExt, TryFutureExt};
+use futures::{FutureExt, TryFutureExt};
 use object_store::path::Path;
 use snafu::{location, Location};
-use std::cmp::Reverse;
+use std::collections::BinaryHeap;
 use std::fmt::Debug;
 use std::future::Future;
 use std::ops::Range;
 use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::{Arc, Mutex};
-use std::time::{Duration, Instant};
-use tokio::sync::{AcquireError, Semaphore, TryAcquireError};
-use tokio::time::error::Elapsed;
+use std::time::Instant;
+use tokio::sync::Notify;
 
 use lance_core::{Error, Result};
 
 use crate::object_store::ObjectStore;
 use crate::traits::Reader;
 
+// Don't log backpressure warnings until at least this many seconds have passed
+const BACKPRESSURE_MIN: u64 = 5;
+// Don't log backpressure warnings more than once / minute
+const BACKPRESSURE_DEBOUNCE: u64 = 60;
+
+// We want to allow requests that have a lower priority than any
+// currently in-flight request.  This helps avoid potential deadlocks
+// related to backpressure.  Unfortunately, it is quite expensive to
+// keep track of which priorities are in-flight.
+//
+// TODO: At some point it would be nice if we can optimize this away but
+// in_flight should remain relatively small (generally less than 256 items)
+// and has not shown itself to be a bottleneck yet.
+struct PrioritiesInFlight {
+    in_flight: Vec<u128>,
```

Comment:
The removal is not necessarily removal from tail (since I/O requests don't necessarily complete in order) so I'm not sure that `VecDeque` would be any faster (O(N) removal in both cases)